README.md
Country-level GeoNames geocoder
===============================

**ggeocode** is a simple geocoding library uses free GeoNames database to construct a fast and simple local Python geocoding library. It is designed to geocode only to the **country** level, and considers only administrative subdivisions and populated places, not other features like parks, rivers, etc. It does not distinguish languages, and will attempt to geocode names in all languages available.

# Installation

    $ cd ggeocode
    $ python setup.py install

# Usage

## One-time preparation

Before using ggeocode, you need to compile your local JSON mapping file from the GeoNames database.

1. Download the GeoNames ``allCountries.zip`` data file from http://download.geonames.org/export/dump/allCountries.zip
2. Unzip the archive to get the file ``allCountries.txt`` (approximately 1.5 GB)
3. Compile a local JSON mapping table using the command below.

Compilation command:

    $ python -m ggeocode.parser allCountries.txt > name-map.lines.json

## Command-line use

You need to have compiled your local JSON mapping file first.  Assuming your mapping file is in ``name-map.lines.json``, you can geocode a collection of names using the following syntax:

    $ python -m ggeocode.coder -s stoplist.txt name-map.lines.json Name1 "Name 2" [... NameN]

The program will print a list of JSON objects (one per line) with the results of geocoding each string, including zero or more ISO3 country codes most-likely associated with each name (the order is not significant). Note that startup time for the command-line version is slow, because it has to read a large database first: the more names you geocode with a single command, the less the overhead will matter.

### Stoplist

There is a sample stoplist in the file ./stoplist.txt, but you can substitute your own. These are common words and phrases that shouldn't be geocoded by themselves (but can be part of a bigger phrase).

### Command-line example

This command-line example attempts to geocode four strings:

    $ python -m ggeocode.coder name-map.lines.json Ottawa "Ottawa KS" Winnipeg Belleville "Belleville, ON"
    {"text": "Ottawa", "status": true, "score": 3, "countries": ["CAN"]}
    {"text": "Ottawa KS", "status": true, "score": 5, "countries": ["USA"]}
    {"text": "Winnipeg", "status": true, "score": 3, "countries": ["CAN"]}
    {"text": "Belleville", "status": true, "score": 2, "countries": ["CAN", "FRA", "USA"]}
    {"text": "Belleville, ON", "status": true, "score": 5, "countries": ["CAN"]}
    $

Note that the only result for "Ottawa" is "CAN" (Canada); because Ottawa is the national capital, ggeocode gives it more prominance, and assumes that's what the user means; however, "Ottawa KS" (Ottawa, Kansas) correctly resolves to "USA". "Winnipeg" is not a national capital, but it's the only city with that name. There are several "Belleville"'s in Canada, France, and the USA, and all get returned as possible results; however, when you specify "Belleville, ON" (Belleville, Ontario), you correctly get only Canada as a result.

## Library usage

To use ggeocode as a library, compile your JSON mapping file as explained in "One-time preparation" above, and include the following at the top of your Python module:

    from ggeocode import coder

You will need to load the JSON mapping file once, and it may take from several seconds to over a minute, depending on your environment:

    coder.load_name_map("name-map.lines.json") # substitute the actual filename

After that, use the _coder.geocode_ function to geocode each string. The return value will be a JSON object, including the following properties:

*text* - the original text being geocoded
*status* - true for success, or false if no matches were found
*score* - the score of the best matches (higher is better; scores below 3 are unreliable)
*countries* - a list of zero or more matching ISO3 country codes (with the best score)

### Library examples

    result = coder.code("Ottawa") # ['CAN']
    result = coder.code("Ottawa KS") # ['USA']
    result = coder.code("Winnipeg") # ['CAN']
    result = coder.code("Belleville") # ['CAN', 'FRA', 'USA']
    result = coder.code("Belleville, ON") # ['CAN']

See "Command-line examples," above, for an explanation of the results.

# License

Started by David Megginson, 2019-09.

This Python module is in the Public Domain. See UNLICENSE.md for details.


CHANGELOG
2020-01-21 Release 0.5:
	- add support for stoplists
	- exclude _ and digits from words (replace with whitespace)
	- add sample stoplist

2019-09-26 Release 0.4:
	- fix bug that could cause mis-coding during long runs
	- add command-line options to the coder for log level, etc.


iso3.py
MAP = {
    "AD": "AND",
    "AE": "ARE",
    "AF": "AFG",
    "AG": "ATG",
    "AI": "AIA",
    "AL": "ALB",
    "AM": "ARM",
    "AO": "AGO",
    "AQ": "ATA",
    "AR": "ARG",
    "AS": "ASM",
    "AT": "AUT",
    "AU": "AUS",
    "AW": "ABW",
    "AX": "ALA",
    "AZ": "AZE",
    "BA": "BIH",
    "BB": "BRB",
    "BD": "BGD",
    "BE": "BEL",
    "BF": "BFA",
    "BG": "BGR",
    "BH": "BHR",
    "BI": "BDI",
    "BJ": "BEN",
    "BL": "BLM",
    "BM": "BMU",
    "BN": "BRN",
    "BO": "BOL",
    "BQ": "BES",
    "BR": "BRA",
    "BS": "BHS",
    "BT": "BTN",
    "BV": "BVT",
    "BW": "BWA",
    "BY": "BLR",
    "BZ": "BLZ",
    "CA": "CAN",
    "CC": "CCK",
    "CD": "COD",
    "CF": "CAF",
    "CG": "COG",
    "CH": "CHE",
    "CI": "CIV",
    "CK": "COK",
    "CL": "CHL",
    "CM": "CMR",
    "CN": "CHN",
    "CO": "COL",
    "CR": "CRI",
    "CU": "CUB",
    "CV": "CPV",
    "CW": "CUW",
    "CX": "CXR",
    "CY": "CYP",
    "CZ": "CZE",
    "DE": "DEU",
    "DJ": "DJI",
    "DK": "DNK",
    "DM": "DMA",
    "DO": "DOM",
    "DZ": "DZA",
    "EC": "ECU",
    "EE": "EST",
    "EG": "EGY",
    "EH": "ESH",
    "ER": "ERI",
    "ES": "ESP",
    "ET": "ETH",
    "FI": "FIN",
    "FJ": "FJI",
    "FK": "FLK",
    "FM": "FSM",
    "FO": "FRO",
    "FR": "FRA",
    "GA": "GAB",
    "GB": "GBR",
    "GD": "GRD",
    "GE": "GEO",
    "GF": "GUF",
    "GG": "GGY",
    "GH": "GHA",
    "GI": "GIB",
    "GL": "GRL",
    "GM": "GMB",
    "GN": "GIN",
    "GP": "GLP",
    "GQ": "GNQ",
    "GR": "GRC",
    "GS": "SGS",
    "GT": "GTM",
    "GU": "GUM",
    "GW": "GNB",
    "GY": "GUY",
    "HK": "HKG",
    "HM": "HMD",
    "HN": "HND",
    "HR": "HRV",
    "HT": "HTI",
    "HU": "HUN",
    "ID": "IDN",
    "IE": "IRL",
    "IL": "ISR",
    "IM": "IMN",
    "IN": "IND",
    "IO": "IOT",
    "IQ": "IRQ",
    "IR": "IRN",
    "IS": "ISL",
    "IT": "ITA",
    "JE": "JEY",
    "JM": "JAM",
    "JO": "JOR",
    "JP": "JPN",
    "KE": "KEN",
    "KG": "KGZ",
    "KH": "KHM",
    "KI": "KIR",
    "KM": "COM",
    "KN": "KNA",
    "KP": "PRK",
    "KR": "KOR",
    "KW": "KWT",
    "KY": "CYM",
    "KZ": "KAZ",
    "LA": "LAO",
    "LB": "LBN",
    "LC": "LCA",
    "LI": "LIE",
    "LK": "LKA",
    "LR": "LBR",
    "LS": "LSO",
    "LT": "LTU",
    "LU": "LUX",
    "LV": "LVA",
    "LY": "LBY",
    "MA": "MAR",
    "MC": "MCO",
    "MD": "MDA",
    "ME": "MNE",
    "MF": "MAF",
    "MG": "MDG",
    "MH": "MHL",
    "MK": "MKD",
    "ML": "MLI",
    "MM": "MMR",
    "MN": "MNG",
    "MO": "MAC",
    "MP": "MNP",
    "MQ": "MTQ",
    "MR": "MRT",
    "MS": "MSR",
    "MT": "MLT",
    "MU": "MUS",
    "MV": "MDV",
    "MW": "MWI",
    "MX": "MEX",
    "MY": "MYS",
    "MZ": "MOZ",
    "NA": "NAM",
    "NC": "NCL",
    "NE": "NER",
    "NF": "NFK",
    "NG": "NGA",
    "NI": "NIC",
    "NL": "NLD",
    "NO": "NOR",
    "NP": "NPL",
    "NR": "NRU",
    "NU": "NIU",
    "NZ": "NZL",
    "OM": "OMN",
    "PA": "PAN",
    "PE": "PER",
    "PF": "PYF",
    "PG": "PNG",
    "PH": "PHL",
    "PK": "PAK",
    "PL": "POL",
    "PM": "SPM",
    "PN": "PCN",
    "PR": "PRI",
    "PS": "PSE",
    "PT": "PRT",
    "PW": "PLW",
    "PY": "PRY",
    "QA": "QAT",
    "RE": "REU",
    "RO": "ROU",
    "RS": "SRB",
    "RU": "RUS",
    "RW": "RWA",
    "SA": "SAU",
    "SB": "SLB",
    "SC": "SYC",
    "SD": "SDN",
    "SE": "SWE",
    "SG": "SGP",
    "SH": "SHN",
    "SI": "SVN",
    "SJ": "SJM",
    "SK": "SVK",
    "SL": "SLE",
    "SM": "SMR",
    "SN": "SEN",
    "SO": "SOM",
    "SR": "SUR",
    "SS": "SSD",
    "ST": "STP",
    "SV": "SLV",
    "SX": "SXM",
    "SY": "SYR",
    "SZ": "SWZ",
    "TC": "TCA",
    "TD": "TCD",
    "TF": "ATF",
    "TG": "TGO",
    "TH": "THA",
    "TJ": "TJK",
    "TK": "TKL",
    "TL": "TLS",
    "TM": "TKM",
    "TN": "TUN",
    "TO": "TON",
    "TR": "TUR",
    "TT": "TTO",
    "TV": "TUV",
    "TW": "TWN",
    "TZ": "TZA",
    "UA": "UKR",
    "UG": "UGA",
    "UM": "UMI",
    "US": "USA",
    "UY": "URY",
    "UZ": "UZB",
    "VA": "VAT",
    "VC": "VCT",
    "VE": "VEN",
    "VG": "VGB",
    "VI": "VIR",
    "VN": "VNM",
    "VU": "VUT",
    "WF": "WLF",
    "WS": "WSM",
    "YE": "YEM",
    "YT": "MYT",
    "ZA": "ZAF",
    "ZM": "ZMB",
    "ZW": "ZWE"
}


geocode.py
import json, logging, re, sys


logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("geocode")

WS_PATTERN = re.compile('\W+')

def normalise (s):
    """ Generate a normalised version of a string """
    return WS_PATTERN.sub(' ', s).lower().strip()

name_map = {}

loaded_count = 0
with open('working/name-country-map.lines.json') as input:
    logger.info("Loading database...")
    for line in input:
        entry = json.loads(line)
        name_map[entry[0]] = entry[1]
        loaded_count += 1
        if (loaded_count % 1000000) == 0:
            logger.info("Read %d entries", loaded_count)

for name in sys.argv[1:]:
    entries = name_map.get(normalise(name))
    max_score = 0
    max_keys = []
    if entries:
        for key in entries:
            if entries[key] > max_score:
                max_keys = [key]
                max_score = entries[key]
            elif entries[key] == max_score:
                max_keys.append(key)
        print(name, max_keys)
    else:
        print("No entries for", name)


parse-geonames.py
""" Compile a name map from GeoNames data

Usage:

    python parse_geonames.py allCountries.txt > name-map.lines.json

The geocode.py module uses the output.

Started 2019-09 by David Megginson
This code is in the public domain
"""

import fileinput, json, logging, re, sys
import iso3
from geocode import normalise


#
# Global variables
#

logger = logging.getLogger("parse-geonames")

keys = (
    "geonameid",
    "name",
    "asciiname",
    "alternatenames",
    "latitude",
    "longitude",
    "feature_class",
    "feature_code",
    "country_code",
    "cc2",
    "admin1_code",
    "admin2_code",
    "admin3_code",
    "admin4_code",
    "population",
    "elevation",
    "dem",
    "timezone",
    "modification_date",
)
""" Column keys for the tab-separated GeoNames data """


FEATURE_WEIGHTS = {
    # countries and country-like things
    'PCL': 3,
    'PCLD': 3,
    'PCLF': 3,
    'PCLI': 3,
    'PCLIX': 3,
    'PCLS': 3,
    # administrative subdivisions
    'ADM1': 2,
    # national or admin1 capital
    'PPLC': 2,
    'PPLA': 2,
}
""" GeoNames feature names that get extra weight """


#
# Main entry point
#

def read_geonames (input=None):
    """ Parse the GeoNames allCountries.txt file into a Python dict.
    Warning: the dict will be large (hundreds of megabytes)
    @param input: the input stream (defaults to standard input)
    @returns: a dict of geo names and country-weight maps
    @see write_geonames
    """
    if input is None:
        input = sys.stdin
        
    mapping_table = {} # the mapping table we'll return
    
    country_codes_seen = set() # keep track of the ones we've seen, for progress reporting

    # Iterate through the tab-separated input text, row by row, and extract the info
    for row in input:

        # Make a dict out of the fields
        record = dict(zip(keys, row.split("\t")))

        # Look up the ISO3 country code
        country_code = iso3.MAP.get(record['country_code'])
        
        if country_code is not None:
            # Extract the alternate names as well as the normative ones
            # (Note: these will be in multiple languages)
            alternate_names = record['alternatenames'].split(",")
            names = [record['name']] + alternate_names

            # We're interested only in administrative subdivisions and populated places
            # (Not parks, rivers, etc)
            if record['feature_class'] in ('A', 'P',):

                # Keep the user updated, since this runs for a while
                if country_code not in country_codes_seen:
                    logger.info("Processing %s", country_code)
                    country_codes_seen.add(country_code)

                # Add all the names to the mapping table with this country code
                for name in names:
                    if name:

                        # lower case and normalise all non-word characters into a single space
                        name = normalise(name)

                        # create the country weight table, if it doesn't already exist
                        if not mapping_table.get(name):
                            mapping_table[name] = dict()

                        # raise to the appropriate weight for the feature type
                        weight = FEATURE_WEIGHTS.get(record['feature_code'], 1) # default weight is 1

                        # add a prominance bonus if there are lots of translations
                        if len(alternate_names) > 5:
                            weight += 1

                        # save the new weight if it's not lower than the existing one
                        current_weight = mapping_table[name].get(country_code, 1)
                        if current_weight <= weight:
                            mapping_table[name][country_code] = weight

    return mapping_table


def write_geonames (mapping_table, output=None):
    """ Write the mapping table out to line-oriented JSON.
    @param mapping_table: the mapping table to write
    @param output: the output stream (defaults to standard output)
    @see read_geonames
    """
    
    if output is None:
        output = sys.stdout
        
    entries_written = 0

    for entry in mapping_table.items():
        json.dump(list(entry), output)
        print('', file=output)
        entries_written += 1
        if (entries_written % 1000000) == 0:
            logger.info("Wrote %d entries", entries_written)


#
# Command-line script entry point: read all files named as arguments and write to standard output
#

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    mapping_table = read_geonames(fileinput.input())
    write_geonames(mapping_table)

# end


ggeocode/parser.py
""" Compile a name map from GeoNames data

Usage:

    python -m ggeocode.parser allCountries.txt > name-map.lines.json

The geocode.py module uses the output.

Started 2019-09 by David Megginson
This code is in the public domain
"""

import fileinput, json, logging, math, re, sys
import ggeocode.iso3
from ggeocode.coder import normalise


#
# Global variables
#

logger = logging.getLogger("parse-geonames")

keys = (
    "geonameid",
    "name",
    "asciiname",
    "alternatenames",
    "latitude",
    "longitude",
    "feature_class",
    "feature_code",
    "country_code",
    "cc2",
    "admin1_code",
    "admin2_code",
    "admin3_code",
    "admin4_code",
    "population",
    "elevation",
    "dem",
    "timezone",
    "modification_date",
)
""" Column keys for the tab-separated GeoNames data """


FEATURE_WEIGHTS = {
    # countries and country-like things
    'PCL': 5,
    'PCLD': 5,
    'PCLF': 5,
    'PCLI': 5,
    'PCLIX': 5,
    'PCLS': 5,
    # administrative subdivisions
    'ADM1': 4,
    # national or admin1 capital
    'PPLC': 3,
    'PPLA': 2,
}
""" GeoNames feature names that get extra weight """


#
# Main entry point
#

def read_geonames (input=None):
    """ Parse the GeoNames allCountries.txt file into a Python dict.
    Warning: the dict will be large (hundreds of megabytes)
    @param input: the input stream (defaults to standard input)
    @returns: a dict of geo names and country-weight maps
    @see write_geonames
    """
    if input is None:
        input = sys.stdin
        
    mapping_table = {} # the mapping table we'll return
    
    country_codes_seen = set() # keep track of the ones we've seen, for progress reporting

    # Iterate through the tab-separated input text, row by row, and extract the info
    for row in input:

        # Make a dict out of the fields
        record = dict(zip(keys, row.split("\t")))

        # Look up the ISO3 country code
        country_code = ggeocode.iso3.MAP.get(record['country_code'])
        
        if country_code is not None:
            # Extract the alternate names as well as the normative ones
            # (Note: these will be in multiple languages)
            alternate_names = record['alternatenames'].split(",")
            names = [record['name']] + alternate_names

            # We're interested only in administrative subdivisions and populated places
            # (Not parks, rivers, etc)
            if record['feature_class'] in ('A', 'P',):

                # Keep the user updated, since this runs for a while
                if country_code not in country_codes_seen:
                    logger.info("Processing %s", country_code)
                    country_codes_seen.add(country_code)

                # Add all the names to the mapping table with this country code
                for name in names:
                    if name:

                        # lower case and normalise all non-word characters into a single space
                        name = normalise(name)

                        # create the country weight table, if it doesn't already exist
                        if not mapping_table.get(name):
                            mapping_table[name] = dict()

                        # raise to the appropriate weight for the feature type
                        weight = FEATURE_WEIGHTS.get(record['feature_code'], 1) # default weight is 1

                        # add a prominance bonus if there are lots of alternate name (1 for every 10 names)
                        weight += math.floor(len(alternate_names) / 8.0)

                        # save the new weight if it's not lower than the existing one
                        current_weight = mapping_table[name].get(country_code, 1)
                        if current_weight <= weight:
                            mapping_table[name][country_code] = weight

    return mapping_table


def write_geonames (mapping_table, output=None):
    """ Write the mapping table out to line-oriented JSON.
    @param mapping_table: the mapping table to write
    @param output: the output stream (defaults to standard output)
    @see read_geonames
    """
    
    if output is None:
        output = sys.stdout
        
    entries_written = 0

    for entry in mapping_table.items():
        json.dump(list(entry), output)
        print('', file=output)
        entries_written += 1
        if (entries_written % 1000000) == 0:
            logger.info("Wrote %d entries", entries_written)


#
# Command-line script entry point: read all files named as arguments and write to standard output
#

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)

    if len(sys.argv) != 2:
        logging.error("Usage: python %s allCountries.txt > name-map.lines.json", sys.argv[0])
        sys.exit(2)

    with open(sys.argv[1]) as input:
        mapping_table = read_geonames(input)
        
    write_geonames(mapping_table, sys.stdout)

# end


ggeocode/coder.py
""" Attempt to associate a country with text (possibly) containing placenames 

Main entry points:

load_name_map(filename)

Load a prepared geonames map from filename. The parse-geonames.py script prepares the map from the GeoNames
allCountries.txt file.

load_stop_list(filename)

Load a list of phrases to ignore, one per line.

code(text, length=0)

Geocode all word sequences in the text up to length words. 
If length==0, geocode the entire text as a single blob.
Returns a (possibly-empty) list of matching ISO3 country codes.

Python usage:

    from ggeocode.coder import load_name_map, code

    load_name_map("name-map.lines.json")
    load_stoplist("stoplist.txt")

    result = code("Ottawa") # ['CAN']
    result = code("Ottawa KS") # ['USA']
    result = code("Belleville") # ['CAN', 'FRA', 'USA']

Command-line usage:

    $ python -m ggeocode.coder name-map.lines.json "Ottawa" "Ottawa KS" "LAX"

Started 2019-09 by David Megginson
This code is in the Public Domain.

"""

import argparse, json, logging, re, readline, sys


#
# Module variables
#

logger = logging.getLogger("geocode")
""" Logger for errors and messages. """

WS_PATTERN = re.compile('[\W\d_]+')
""" Precompiled regular expression for non-word characters. """

name_map = {}
""" Compiled map from GeoNames data """

stoplist = set()
""" List of normalised phrases to ignore """


#
# Utility functions
#

def normalise (s):
    """ Generate a normalised version of a string.
    @param s: the string to normalise
    @returns: the normalise string
    """
    return WS_PATTERN.sub(' ', s).lower().strip()


def merge_weight_map(main_map, merge_map, num_words):
    """ Merge a new weight map into an existing one.
    If the same key exists in both maps, add the values.
    @param main_map: the existing weight map (may be None)
    @param merge_map: the new weight map to merge (may be None)
    @param num_words: the number of words in the phrase (more words == higher)
    @returns: the union of the two maps, with values added.
    """

    # weighting formula for multi-word phrases
    factor = 1 + ((num_words - 1) * 2)
    logger.debug("%d words, factor %f", num_words, factor)
    
    if main_map is None:
        main_map = dict()

    for key in merge_map:
        if key in main_map:
            main_map[key] += merge_map[key] * factor
        else:
            main_map[key] = merge_map[key] * factor

    return main_map


def make_result (text, weight_map):
    """ Extract the (possibly-empty) list of country codes with the highest weight.
    Order is not significant.
    @param text: the original text
    @param weight_map: a weight map (country codes as keys, weights as values)
    @returns: a list of country codes (may be empty)
    """
    max_keys = []
    max_score = 0
    if weight_map:
        for key in weight_map:
            if weight_map[key] > max_score:
                max_keys = [key]
                max_score = weight_map[key]
            elif weight_map[key] == max_score:
                max_keys.append(key)
    return {
        "text": text,
        "status": True if max_score > 0 else False,
        "score": max_score,
        "countries": max_keys,
    }
        
#
# External entry points
#

def load_name_map (filename):
    """ Load the pre-compiled GeoNames map.
    The map is line-oriented JSON. Each line contains a JSON array with a normalised name as
    the first element, and a weight map as the second element. Use the parse-geonames.py
    script to create the file from the GeoNames allCountries.txt file.
    Will load the values into the module-global name_map dictionary.
    @param filename: the filename containing the line-oriented JSON.
    @see name_map
    """
    if not name_map:
        loaded_count = 0
        with open(filename) as input:
            logger.info("Loading database...")
            for line in input:
                entry = json.loads(line)
                name_map[entry[0]] = entry[1]
                loaded_count += 1
                if (loaded_count % 1000000) == 0:
                    logger.info("Read %d entries", loaded_count)

def load_stoplist(filename):
    """ Load a stoplist of words and phrases, one per line.
    Any phrase in this list will not be geocoded.
    @param filename: the filename containing the stop phrases
    @see stoplist
    """
    with open(filename) as input:
        logger.info("Loading stoplist...")
        for line in input:
            line = re.sub('#.*$', '', line)
            line = normalise(line)
            if line:
                stoplist.add(line)


def code (text, max_words=0):
    """ Return the most-likely list of country codes for a text string.
    You must call load_name_map() before invoking this function.
    Will normalise the text, then attempt to geocode multi-word phrases up to
    max_words words. If max_words is 0, then try to geocode all of the text as a 
    single string.
    For example, if max_words is 3 and the text string is "In Los Angeles, California"
    this function will try to geocode the following strings:
    "in los angeles"
    "los angeles california"
    "in los"
    "los angeles"
    "angeles california"
    "in"
    "los"
    "angeles"
    "california"
    @param text: the text string to geocode
    @param max_words: the maximum number of words in each phrase
    @returns: a (possibly-empty) list of country codes with the highest weight
    @see load_name_map
    """
    if not name_map:
        raise Exception("No name map loaded. You must call load_name_map(filename) first")

    # simplify the text (remove extra spaces and punctuation, and make lower case)
    text_norm = normalise(text)

    if max_words == 0:
        # 0 means just try to geocode the whole thing as a single string
        weight_map = name_map.get(text_norm)
    else:
        # an integer >0 means try variable-length phrases
        weight_map = None # map that will hold the merged results

        words = text_norm.split(" ") # make an array of words

        # test all phrases >= max_words words long
        for i in range(max_words, 0, -1):
            for j in range(0, len(words) - i + 1):
                # rejoin the phrase as a string
                key = " ".join(words[j:j+i])

                # skip single letters or phrases in the stoplist
                if len(key) == 1 or key in stoplist:
                    continue

                logger.debug("Trying %s...", key)
                result = name_map.get(key)
                if result:
                    logger.debug("Match: %s", result)
                    # merge into the combined weight map
                    weight_map = merge_weight_map(weight_map, result, i)
                    logger.debug("Current weight map: %s", str(weight_map))

    # return a (possibly-empty) list of the country codes with the highest weight
    return make_result(text, weight_map)


#
# Command-line script entry point: try to geocode all the arguments, or read strings from the console.
#

if __name__ == "__main__":

    # process args
    arg_parser = argparse.ArgumentParser(description="Geocode text (by country)")
    arg_parser.add_argument("-m", "--name-map", required=True, help="File containing pre-compiled name map.")
    arg_parser.add_argument("-w", "--max-words", type=int, default=3, help="Maximum phrase length to geocode (0 means do the whole string).")
    arg_parser.add_argument("-s", "--stoplist", help="File containing phrases to ignore (one per line).")
    arg_parser.add_argument(
        "-l", "--log-level",
        default="info",
        choices={"debug", "info", "warning", "error"},
        help="Level for logging messages (default: info)."
    )
    arg_parser.add_argument("strings", nargs="*", help="Text to geocode (if non supplied, read strings from command-line.")
    args = arg_parser.parse_args();

    # set up logging
    logging.basicConfig(level=args.log_level.upper())

    # load the pre-compiled map
    load_name_map(args.name_map)

    # load stoplist
    if args.stoplist:
        load_stoplist(args.stoplist)

    # geocode
    if len(args.strings) > 0:
        # We have strings to geocode on the command line
        for s in args.strings:
            print(json.dumps(code(s, max_words=args.max_words)))
    else:
        # We'll read input interactively from the console
        def read_input ():
            try:
                s = input("ggeocode> ")
                return s.strip().lower()
            except EOFError:
                print("Quitting...")
                sys.exit(0)

        while True:
            # Endless console loop
            s = read_input()
            if s == "exit":
                sys.exit(0)
            elif s in ("?", "h", "help"):
                print("Type a string you'd like to geocode (\"exit\" or Ctrl-D to exit)")
            else:
                print(json.dumps(code(s, max_words=args.max_words)))

# end
